File: apps/api/app/schemas/rickandmorty.py
from enum import Enum

from pydantic import BaseModel


class CharacterStatus(str, Enum):
    ALIVE = "Alive"
    DEAD = "Dead"
    UNKNOWN = "unknown"


class CharacterGender(str, Enum):
    MALE = "Male"
    FEMALE = "Female"
    GENDERLESS = "Genderless"
    UNKNOWN = "unknown"


class RickAndMortyCharacterRequest(BaseModel):
    name: str | None = None
    status: CharacterStatus | None = None
    species: str | None = None
    type: str | None = None
    gender: CharacterGender | None = None
    page: int | None = None


class RickAndMortyCharacter(BaseModel):
    id: int
    name: str | None = None
    status: CharacterStatus | None = None
    species: str | None = None
    type: str | None = None
    gender: CharacterGender | None = None
    origin: dict | None = None
    location: dict | None = None
    image: str | None = None
    episode: list[str] | None = None
    url: str | None = None
    created: str | None = None


class RickAndMortyPagination(BaseModel):
    count: int
    pages: int
    next: str | None = None
    prev: str | None = None


class RickAndMortyCharacterResponse(BaseModel):
    info: RickAndMortyPagination
    results: list[RickAndMortyCharacter] | None = None


class RickAndMortyLocationRequest(BaseModel):
    name: str | None = None
    type: str | None = None
    dimension: str | None = None


class RickAndMortyLocation(BaseModel):
    id: int
    name: str | None = None
    type: str | None = None
    dimension: str | None = None
    residents: list[str] | None = None
    url: str | None = None
    created: str | None = None


class RickAndMortyLocationResponse(BaseModel):
    info: RickAndMortyPagination
    results: list[RickAndMortyLocation] | None = None


class RickAndMortyEpisodeRequest(BaseModel):
    name: str | None = None
    episode: str | None = None


class RickAndMortyEpisode(BaseModel):
    id: int
    name: str | None = None
    air_date: str | None = None
    episode: str | None = None
    characters: list[str] | None = None
    url: str | None = None
    created: str | None = None


class RickAndMortyEpisodeResponse(BaseModel):
    info: RickAndMortyPagination
    results: list[RickAndMortyEpisode] | None = None

File: apps/web/src/types/rickandmorty.ts
export enum CharacterStatus {
  ALIVE = 'Alive',
  DEAD = 'Dead',
  UNKNOWN = 'unknown',
}

export enum CharacterGender {
  MALE = 'Male',
  FEMALE = 'Female',
  GENDERLESS = 'Genderless',
  UNKNOWN = 'unknown',
}

export type CharacterStatusValue = `${CharacterStatus}`
export type CharacterGenderValue = `${CharacterGender}`

export interface RickAndMortyCharacterRequest {
  name?: string | null
  status?: CharacterStatusValue | null
  species?: string | null
  type?: string | null
  gender?: CharacterGenderValue | null
  page?: number | null
  ids?: number[] | null
}

export interface LocationReference {
  name: string
  url: string
}

export interface RickAndMortyCharacter {
  id: number
  name?: string | null
  status?: CharacterStatusValue | null
  species?: string | null
  type?: string | null
  gender?: CharacterGenderValue | null
  origin?: LocationReference | null
  location?: LocationReference | null
  image?: string | null
  episode?: string[] | null
  url?: string | null
  created?: string | null
}

export interface RickAndMortyPagination {
  count: number
  pages: number
  next?: string | null
  prev?: string | null
}

export interface RickAndMortyCharacterResponse {
  info: RickAndMortyPagination
  results?: RickAndMortyCharacter[] | null
}

export interface RickAndMortyLocationRequest {
  name?: string | null
  type?: string | null
  dimension?: string | null
}

export interface RickAndMortyLocation {
  id: number
  name?: string | null
  type?: string | null
  dimension?: string | null
  residents?: string[] | null
  url?: string | null
  created?: string | null
}

export interface RickAndMortyLocationResponse {
  info: RickAndMortyPagination
  results?: RickAndMortyLocation[] | null
}

export interface RickAndMortyEpisodeRequest {
  name?: string | null
  episode?: string | null
}

export interface RickAndMortyEpisode {
  id: number
  name?: string | null
  air_date?: string | null
  episode?: string | null
  characters?: string[] | null
  url?: string | null
  created?: string | null
}

export interface RickAndMortyEpisodeResponse {
  info: RickAndMortyPagination
  results?: RickAndMortyEpisode[] | null
}
